### ISSUE TITLE
tex.catwitherror: Wrong number of digits

### ISSUE THREAD

**Issue by martin-ueding** [HISKP-LQCD/hadron#283]
I have found an example where the function gives a misleading result. My value is 0.0008970, the error is 0.0002106. So the error has the same order as the value, but it is smaller. When giving no `digits` argument, the output is fine.

```
> hadron::tex.catwitherror(0.0008970, 0.0002106, with.dollar = FALSE)
[1] "9(2)\\cdot 10^{-04}"
```

But with two digits the output is just wrong. It makes it seem that the error is much larger than the value.

```
> hadron::tex.catwitherror(0.0008970, 0.0002106, with.dollar = FALSE, digits = 2)
[1] "9(21)\\cdot 10^{-04}"
```

Adding another digit makes it better again.

```
> hadron::tex.catwitherror(0.0008970, 0.0002106, with.dollar = FALSE, digits = 3)
[1] "0.000897(211)"
```

I am not sure whether that only happens on this particular magnitude, but it potentially means that results are wrong.

**Comment by kostrzewa** [comment]
Is this definitely with a hadron version which is up to date with the master branch?

**Comment by martin-ueding** [comment]
Yes, I apparently have used the latest version for this test. Can't you reproduce it?

**Comment by kostrzewa** [comment]
I hadn't tried to reproduce it, just wanted to make sure as some effort was invested by @j-ostmeyer to fix these various issues. I guess we should write another test and try to fix this behaviour... The logic tree for `tex.catwitherror` is unfortunately rather complicated (somewhat by necessity).

**Comment by kostrzewa** [comment]
ps: I reproduced it now

**Comment by j-ostmeyer** [comment]
That bug is due to our trust into R-base functions.
```
> format(round(0.000897, digits=4), nsmall=4)
[1] "9e-04"
> format(round(0.000897, digits=5), nsmall=5)
[1] "9e-04"
```
The `nsmall` doesn't fulfill its purpose here.

**Comment by j-ostmeyer** [comment]
The plain version without scientific notation works fine:
```
> tex.catwitherror(0.0008970, 0.0002106, with.dollar = FALSE, digits = 2, human.readable=FALSE)
[1] "0.00090(21)"
```

**Comment by j-ostmeyer** [comment]
Maybe we should just switch to some package specialized on this...
The `errors`-package gives:
```
> format(set_errors(0.0008970, 0.0002106), digits = 2)
[1] "9.0(21)e-4"
```
This is not _precisely_ what we would like to have, but the notation is consistent and I believe that the implementation is more stable than ours. We can then of course still add things like dollars and so on.

**Comment by urbach** [comment]
Knitr also has this capability, but I don't know how to use it directly. It is automatically used in documents...
-- 
CU Mobil, www.carsten-urbach.de

**Comment by martin-ueding** [comment]
Why isn't `9.0(21)e-4` exactly what we want for the non-latex case?

**Comment by j-ostmeyer** [comment]
In the way we have it I would expect `9.0(2.1)e-4`. But as I said, the form above is fine, too.

**Comment by urbach** [comment]
hmm, somehow `tex.catwitherror` has not become more reliable over time. 

    > tex.catwitherror(x=1.924400e+01, dx=1.348052e+00, digits=2)
    [1] "$19.2(13)$"
    > tex.catwitherror(x=1.613799e+00, dx=3.404626e+00, digits=2)
    [1] "$1.6(3.4)$"
    tex.catwitherror(x=5.939860e+13, dx=3.924350e+08, digits=2)
    [1] "$5.93986(392435000)\\cdot 10^{+13}$"

I might remeber wrongly, but I thought that some of these things worked better in the past, but I'm not sure, of course...

**Comment by j-ostmeyer** [comment]
Might be. I cleaned it up recently and focused on the examples we implemented in `testthat`.  If some edge cases have not been included in these examples and I did not think of them, they might well have become worse. One reason more to switch to some more reliable package like `errors`.

**Comment by urbach** [comment]
> Might be. I cleaned it up recently and focused on the examples we implemented in `testthat`.  If some edge cases have not been included in these examples and I did not think of them, they might well have become worse. One reason more to switch to some more reliable package like `errors`.

could we map most of our functionality to errors?

**Comment by j-ostmeyer** [comment]
Let me try tomorrow. I believe that should be possible and not very complicated.

**Comment by j-ostmeyer** [comment]
Shall I create the pull request to `forCran` or to `master`?

**Comment by urbach** [comment]
> Shall I create the pull request to `forCran` or to `master`?

master. This will not go into the CRAN version right now, unless there
are more modifications requested.

**Comment by j-ostmeyer** [comment]
See #285
